From b7ba36062a85cc06cb9c562e12611215795f3663 Mon Sep 17 00:00:00 2001
From: Pierre Riteau <priteau@uchicago.edu>
Date: Mon, 10 Jul 2017 16:05:19 +0100
Subject: [PATCH] Support LDAP-hashed passwords

---
 keystone/common/utils.py | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/keystone/common/utils.py b/keystone/common/utils.py
index 2a2cdbff6..0fb0eaad9 100644
--- a/keystone/common/utils.py
+++ b/keystone/common/utils.py
@@ -146,7 +146,12 @@ def check_password(password, hashed):
     if password is None or hashed is None:
         return False
     password_utf8 = verify_length_and_trunc_password(password).encode('utf-8')
-    return passlib.hash.sha512_crypt.verify(password_utf8, hashed)
+    if hashed.startswith('{CRYPT}'):
+        return passlib.hash.ldap_md5_crypt.verify(password_utf8, hashed)
+    elif hashed.startswith('{MD5}'):
+        return passlib.hash.ldap_md5.verify(password_utf8, hashed)
+    else:
+        return passlib.hash.sha512_crypt.verify(password_utf8, hashed)
 
 
 def attr_as_boolean(val_attr):
-- 
2.13.2

